Detectors/GlobalTracking/src/MatchTOF.cxx

```diff
@@ -1443,22 +1556,55 @@ void MatchTOF::BestMatches(std::vector<o2::dataformats::MatchInfoTOFReco>& match
 
     int itrk = matchingPair.getIdLocal();
 
-    if (matchedTracksIndex[trkType][itrk] != -1) { // the track was already filled
-      continue;
-    }
-    if (matchedClustersIndex[matchingPair.getTOFClIndex()] != -1) { // the cluster was already filled
-      continue;
-    }
-    matchedTracksIndex[trkType][itrk] = matchedTracks[trkType].size();                      // index of the MatchInfoTOF correspoding to this track
-    matchedClustersIndex[matchingPair.getTOFClIndex()] = matchedTracksIndex[trkType][itrk]; // index of the track that was matched to this cluster
-
     int trkTypeSplitted = trkType;
     auto sourceID = matchingPair.getTrackRef().getSource();
     if (sourceID == o2::dataformats::GlobalTrackID::TPCTRD) {
       trkTypeSplitted = (int)trkType::TPCTRD;
     } else if (sourceID == o2::dataformats::GlobalTrackID::ITSTPCTRD) {
       trkTypeSplitted = (int)trkType::ITSTPCTRD;
     }
+
+    if (matchedClustersIndex[matchingPair.getTOFClIndex()] != -1) { // the cluster was already filled
+      continue;
+    }
+    if (matchedTracksIndex[trkType][itrk] != -1) { // the track was already filled
+      if (trkType) {                               // not applied for unconstrained tracks (trkType == 0)
+        // let's check if we can use both matching to improve TOF time
+        int pairInd = matchedTracksIndex[trkType][itrk];
+        auto& prevMatching = matchedTracks[trkTypeSplitted][pairInd];
+
+        if (pairInd >= matchedTracks[trkTypeSplitted].size()) {
+          LOG(error) << "Pair index out of range when trying to update TOF time. This should not happen";
+          continue;
+        }
+
+        int istripOld = TOFClusWork[prevMatching.getTOFClIndex()].getMainContributingChannel() / o2::tof::Geo::NPADS;
+        int istripNew = TOFClusWork[matchingPair.getTOFClIndex()].getMainContributingChannel() / o2::tof::Geo::NPADS;
+
+        if (istripOld != istripNew) { // we accept it only if in a different strip
+          const o2::track::TrackLTIntegral& intLTnew = matchingPair.getLTIntegralOut();
+          const o2::track::TrackLTIntegral& intLTold = prevMatching.getLTIntegralOut();
+
+          const float cinv = 1.E+10 / TMath::C(); // cm/ps
+          float deltaT = (intLTnew.getL() - intLTold.getL()) * cinv;
+
+          float timeNew = TOFClusWork[matchingPair.getTOFClIndex()].getTime() - deltaT;
+          float timeOld = TOFClusWork[prevMatching.getTOFClIndex()].getTime();
+
+          if (fabs(timeNew - timeOld) < 200) {
+            // update time information averaging the two (the second one corrected for the difference in the track length)
+            prevMatching.setSignal((timeNew + timeOld) * 0.5);
+            prevMatching.setChi2(0);                                                                // flag such cases with chi2 equal to zero
+            matchedClustersIndex[matchingPair.getTOFClIndex()] = matchedTracksIndex[trkType][itrk]; // flag also the second cluster as already used
+          }
```

Comment:
Not sure: what will be the cluster reference in the matchInfo in that case? Is the info about the 2nd cluster lost?

Reply:
the first cluster matched will be propagated as info in the matchInfo, the latter one is lost (only used to refine time info). The current matchInfo is not able to trace both. Note that in the previous version the second one was discarded. In any case this information will be lost in AO2D.
For debug studies using matchable clusters, all infos are kept since the output stored is the one before of the BestMaches call.

Reply:
OK, but currently global tracking extracts the track time directly from the associated cluster rather than from the `MatchInfoTOF.mSignal`, so with the double cluster match there will a discrepancy, see
https://github.com/AliceO2Group/AliceO2/blob/2f440e53a5c41f6c748427a54cab18774f688573/DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h#L153-L154

https://github.com/AliceO2Group/AliceO2/blob/2f440e53a5c41f6c748427a54cab18774f688573/DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h#L175-L176

https://github.com/AliceO2Group/AliceO2/blob/2f440e53a5c41f6c748427a54cab18774f688573/DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h#L226-L227

and also 
https://github.com/AliceO2Group/AliceO2/blob/2f440e53a5c41f6c748427a54cab18774f688573/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx#L1494-L1495

https://github.com/AliceO2Group/AliceO2/blob/2f440e53a5c41f6c748427a54cab18774f688573/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx#L1504-L1505

https://github.com/AliceO2Group/AliceO2/blob/2f440e53a5c41f6c748427a54cab18774f688573/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx#L1514-L1515

Could you then modify this time assignment to use mSignal?

Reply:
You are right. I overlooked it. Yes I can do it.

Reply:
Done.
For your info, it seems to me that now tofCl is no longer used in data both in RecoContainer and CreateVariadic but it is needed for MC.